deepspeed/inference/v2/kernels/ragged_ops/moe_scatter/moe_scatter.cu

```diff
@@ -15,10 +15,11 @@ namespace scatter {
 constexpr int access_granularity = 16;
 constexpr int threads = 256;
 constexpr int warps = threads / hw_warp_size;
+constexpr int max_experts = 1024;
 
 }  // namespace scatter
 
-template <typename T, int copyUnroll>
+template <typename T, int copyUnroll, int N_TOP_K>
```

Comment:
@cmikeh2, I know you like to generalize this function, but I was wondering if we can have two kernels here, one for top-1 and one for top-k, just so that we can remove some of the complexity added for top-1. what do u think?

Reply:
Are you observing any slowdown with top-1? The re-org intention here was primarily to simplify things. Previously, on block-0 we did a cumsum for the GEMM kernel and the rest of the thread blocks did max reductions. The max reduction is of similar complexity to the cumsum anyways (`log(n)`) steps and since it's necessary on all blocks for the top-N case anyways, I thought it made sense to remove the branch from the code and have a unified path.

Reply:
i haven't but i will try to do some profiling of this in the next days. thanks for the clarification on the changes :)